File: unpruuf/app/src/main/java/com/nexonai/unpruuf/domain/network/ratchet/DoubleRatchet.kt
package com.nexonai.unpruuf.domain.network.ratchet

import com.google.crypto.tink.subtle.Hkdf
import com.google.crypto.tink.subtle.X25519
import com.google.crypto.tink.subtle.XChaCha20Poly1305
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Textbook Double Ratchet (Signal spec: https://signal.org/docs/specifications/doubleratchet/),
 * built directly on Tink's `subtle` primitives (X25519, HKDF, XChaCha20-Poly1305) so it needs
 * no extra native dependency and no Android framework class — every symbol here is plain
 * JVM/Kotlin, which keeps it runnable under a regular JUnit test.
 *
 * This file is deliberately app-agnostic: it knows nothing about contacts, pairing or QR
 * codes. Bootstrapping the very first root key from the unpruuf pairing handshake is
 * [com.nexonai.unpruuf.domain.network.ratchet.RatchetSessionManager]'s job, not this one's.
 */
object DoubleRatchet {

    /** Bounded so a malicious/broken peer can't force unbounded memory growth. */
    const val MAX_SKIPPED_KEYS = 1000

    private const val DH_KEY_LEN = 32
    private const val ROOT_KDF_INFO = "unpruuf-ratchet-root-v1"
    private const val CHAIN_KDF_MAC = "HmacSha256"

    class RatchetError(message: String) : Exception(message)

    data class KeyPair(val privateKey: ByteArray, val publicKey: ByteArray)

    /**
     * Mutable ratchet session state for one direction of one contact. All fields are
     * intentionally `var`/mutable — [encrypt] and [decrypt] advance them in place, and the
     * caller ([RatchetSessionManager]) is responsible for persisting the result after every
     * call so forward secrecy survives an app restart (old keys are simply overwritten).
     */
    class RatchetState(
        var dhsPrivateKey: ByteArray,
        var dhsPublicKey: ByteArray,
        var dhr: ByteArray?,
        var rootKey: ByteArray,
        var sendChainKey: ByteArray?,
        var recvChainKey: ByteArray?,
        var sendCount: Int = 0,
        var recvCount: Int = 0,
        var previousChainLength: Int = 0,
        val skippedKeys: MutableMap<String, ByteArray> = mutableMapOf(),
    )

    data class EncryptedMessage(val header: RatchetHeader, val ciphertext: ByteArray)

    fun generateKeyPair(): KeyPair {
        val priv = X25519.generatePrivateKey()
        return KeyPair(priv, X25519.publicFromPrivate(priv))
    }

    /**
     * Initialise as the party that already knows the peer's ratchet public key (the QR
     * scanner role in [com.nexonai.unpruuf.screens.qrpair.QrPairViewModel]). Performs the
     * initial sending DH step immediately, so this side can send right away.
     */
    fun initSender(sharedRootKey: ByteArray, ownKeyPair: KeyPair, theirPublicKey: ByteArray): RatchetState {
        val state = RatchetState(
            dhsPrivateKey = ownKeyPair.privateKey,
            dhsPublicKey = ownKeyPair.publicKey,
            dhr = theirPublicKey,
            rootKey = sharedRootKey,
            sendChainKey = null,
            recvChainKey = null,
        )
        val (rk, ck) = kdfRootKey(state.rootKey, dh(state.dhsPrivateKey, theirPublicKey))
        state.rootKey = rk
        state.sendChainKey = ck
        return state
    }

    /**
     * Initialise as the party that does NOT yet know the peer's ratchet public key (the QR
     * "initiator" role, before the partner's first message arrives). Cannot send until
     * [decrypt] has processed at least one inbound message — that is correct Double Ratchet
     * behaviour, not a bug: only [decrypt] discovers [theirPublicKey] and completes the
     * matching sending chain via the standard DH-ratchet step.
     */
    fun initReceiver(sharedRootKey: ByteArray, ownKeyPair: KeyPair): RatchetState =
        RatchetState(
            dhsPrivateKey = ownKeyPair.privateKey,
            dhsPublicKey = ownKeyPair.publicKey,
            dhr = null,
            rootKey = sharedRootKey,
            sendChainKey = null,
            recvChainKey = null,
        )

    fun encrypt(state: RatchetState, plaintext: ByteArray, associatedData: ByteArray): EncryptedMessage {
        val chainKey = state.sendChainKey
            ?: throw RatchetError("no sending chain yet — must receive at least one message first")
        val (nextChainKey, messageKey) = kdfChainKey(chainKey)
        state.sendChainKey = nextChainKey
        val header = RatchetHeader(state.dhsPublicKey, state.previousChainLength, state.sendCount)
        state.sendCount += 1
        val ciphertext = aead(messageKey).encrypt(plaintext, associatedData + header.encode())
        return EncryptedMessage(header, ciphertext)
    }

    fun decrypt(state: RatchetState, header: RatchetHeader, ciphertext: ByteArray, associatedData: ByteArray): ByteArray {
        val fullAd = associatedData + header.encode()

        trySkippedKey(state, header)?.let { messageKey ->
            return aead(messageKey).decrypt(ciphertext, fullAd)
        }

        if (state.dhr == null || !header.dhPub.contentEquals(state.dhr)) {
            skipMessageKeys(state, header.previousChainLength)
            dhRatchetStep(state, header.dhPub)
        }

        skipMessageKeys(state, header.messageNumber)
        val chainKey = state.recvChainKey ?: throw RatchetError("receiving chain not established")
        val (nextChainKey, messageKey) = kdfChainKey(chainKey)
        state.recvChainKey = nextChainKey
        state.recvCount += 1
        return aead(messageKey).decrypt(ciphertext, fullAd)
    }

    // ---- DH ratchet ---------------------------------------------------------

    private fun dhRatchetStep(state: RatchetState, theirNewPublicKey: ByteArray) {
        state.previousChainLength = state.sendCount
        state.sendCount = 0
        state.recvCount = 0
        state.dhr = theirNewPublicKey

        val (rk1, recvChain) = kdfRootKey(state.rootKey, dh(state.dhsPrivateKey, theirNewPublicKey))
        state.rootKey = rk1
        state.recvChainKey = recvChain

        val newKeyPair = generateKeyPair()
        state.dhsPrivateKey = newKeyPair.privateKey
        state.dhsPublicKey = newKeyPair.publicKey

        val (rk2, sendChain) = kdfRootKey(state.rootKey, dh(state.dhsPrivateKey, theirNewPublicKey))
        state.rootKey = rk2
        state.sendChainKey = sendChain
    }

    // ---- Skipped message keys (out-of-order delivery) ------------------------

    private fun skipMessageKeys(state: RatchetState, until: Int) {
        val chainKey = state.recvChainKey ?: return // nothing received on this chain yet
        if (until - state.recvCount > MAX_SKIPPED_KEYS) {
            throw RatchetError("too many skipped messages (${until - state.recvCount})")
        }
        var ck = chainKey
        while (state.recvCount < until) {
            val (nextChainKey, messageKey) = kdfChainKey(ck)
            ck = nextChainKey
            state.skippedKeys[skippedKeyId(state.dhr!!, state.recvCount)] = messageKey
            if (state.skippedKeys.size > MAX_SKIPPED_KEYS) {
                throw RatchetError("skipped-key cache overflow")
            }
            state.recvCount += 1
        }
        state.recvChainKey = ck
    }

    private fun trySkippedKey(state: RatchetState, header: RatchetHeader): ByteArray? {
        val id = skippedKeyId(header.dhPub, header.messageNumber)
        return state.skippedKeys.remove(id)
    }

    private fun skippedKeyId(dhPub: ByteArray, n: Int): String =
        Base64.getEncoder().encodeToString(dhPub) + ":" + n

    // ---- Primitives -----------------------------------------------------------

    private fun dh(privateKey: ByteArray, publicKey: ByteArray): ByteArray =
        X25519.computeSharedSecret(privateKey, publicKey)

    /** KDF_RK: derives a fresh root key + chain key from the running root key and a new DH output. */
    private fun kdfRootKey(rootKey: ByteArray, dhOutput: ByteArray): Pair<ByteArray, ByteArray> {
        val out = Hkdf.computeHkdf(
            "HmacSha256",
            dhOutput,
            rootKey,
            ROOT_KDF_INFO.toByteArray(Charsets.UTF_8),
            64,
        )
        return out.copyOfRange(0, 32) to out.copyOfRange(32, 64)
    }

    /** KDF_CK: advances a symmetric chain by one step, yielding (nextChainKey, messageKey). */
    private fun kdfChainKey(chainKey: ByteArray): Pair<ByteArray, ByteArray> {
        val nextChainKey = hmac(chainKey, byteArrayOf(0x02))
        val messageKey = hmac(chainKey, byteArrayOf(0x01))
        return nextChainKey to messageKey
    }

    private fun hmac(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance(CHAIN_KDF_MAC)
        mac.init(SecretKeySpec(key, CHAIN_KDF_MAC))
        return mac.doFinal(data)
    }

    private fun aead(messageKey: ByteArray) = XChaCha20Poly1305(messageKey)
}

File: unpruuf/ios/UnpruufCore/Sources/UnpruufCore/Ratchet/DoubleRatchet.swift
import CryptoKit
import CryptoSwift
import Foundation

/// Textbook Double Ratchet (Signal spec: https://signal.org/docs/specifications/doubleratchet/),
/// a direct port of the Android app's `DoubleRatchet.kt`. Built on CryptoKit for every primitive
/// it has natively (X25519, HKDF, HMAC-SHA256) plus CryptoSwift for the one primitive CryptoKit
/// lacks: XChaCha20-Poly1305 (CryptoKit's `ChaChaPoly` is the 12-byte-nonce IETF variant, not the
/// 24-byte-nonce XChaCha20 variant the Android side uses via Tink). All CryptoSwift-specific
/// calls are isolated in the `aeadSeal`/`aeadOpen` pair at the bottom of this file — if
/// CryptoSwift's exact API surface doesn't match what's assumed here, the fix is contained to
/// those two functions and does not touch the ratchet algorithm itself.
///
/// **Wire compatibility requirement:** the AEAD output layout here — `nonce(24) || ciphertext ||
/// tag(16)` — must exactly match Tink's `XChaCha20Poly1305.encrypt()` on the Android side (see
/// that class's doc comment). This is the single highest-value thing to double-check first if
/// cross-platform messages fail to decrypt.
public enum DoubleRatchet {

    /// Bounded so a malicious/broken peer can't force unbounded memory growth.
    public static let maxSkippedKeys = 1000

    private static let rootKdfInfo = "unpruuf-ratchet-root-v1".data(using: .utf8)!

    public struct RatchetError: Error {
        public let message: String
        public init(_ message: String) { self.message = message }
    }

    public struct KeyPair {
        public let privateKey: Data
        public let publicKey: Data
        public init(privateKey: Data, publicKey: Data) {
            self.privateKey = privateKey
            self.publicKey = publicKey
        }
    }

    /// Mutable ratchet session state for one direction of one contact. All fields are
    /// intentionally mutable — `encrypt`/`decrypt` advance them in place, and the caller is
    /// responsible for persisting the result after every call so forward secrecy survives an app
    /// restart (old keys are simply overwritten).
    public final class RatchetState {
        public var dhsPrivateKey: Data
        public var dhsPublicKey: Data
        public var dhr: Data?
        public var rootKey: Data
        public var sendChainKey: Data?
        public var recvChainKey: Data?
        public var sendCount: Int32
        public var recvCount: Int32
        public var previousChainLength: Int32
        public var skippedKeys: [String: Data]

        public init(
            dhsPrivateKey: Data, dhsPublicKey: Data, dhr: Data?, rootKey: Data,
            sendChainKey: Data?, recvChainKey: Data?, sendCount: Int32 = 0, recvCount: Int32 = 0,
            previousChainLength: Int32 = 0, skippedKeys: [String: Data] = [:]
        ) {
            self.dhsPrivateKey = dhsPrivateKey
            self.dhsPublicKey = dhsPublicKey
            self.dhr = dhr
            self.rootKey = rootKey
            self.sendChainKey = sendChainKey
            self.recvChainKey = recvChainKey
            self.sendCount = sendCount
            self.recvCount = recvCount
            self.previousChainLength = previousChainLength
            self.skippedKeys = skippedKeys
        }
    }

    public struct EncryptedMessage {
        public let header: RatchetHeader
        public let ciphertext: Data
    }

    public static func generateKeyPair() -> KeyPair {
        let priv = Curve25519.KeyAgreement.PrivateKey()
        return KeyPair(privateKey: priv.rawRepresentation, publicKey: priv.publicKey.rawRepresentation)
    }

    /// Initialise as the party that already knows the peer's ratchet public key. Performs the
    /// initial sending DH step immediately, so this side can send right away.
    public static func initSender(sharedRootKey: Data, ownKeyPair: KeyPair, theirPublicKey: Data) throws -> RatchetState {
        let state = RatchetState(
            dhsPrivateKey: ownKeyPair.privateKey, dhsPublicKey: ownKeyPair.publicKey,
            dhr: theirPublicKey, rootKey: sharedRootKey, sendChainKey: nil, recvChainKey: nil
        )
        let (rk, ck) = try kdfRootKey(rootKey: state.rootKey, dhOutput: try dh(privateKey: state.dhsPrivateKey, publicKey: theirPublicKey))
        state.rootKey = rk
        state.sendChainKey = ck
        return state
    }

    /// Initialise as the party that does NOT yet know the peer's ratchet public key. Cannot send
    /// until `decrypt` has processed at least one inbound message — that is correct Double
    /// Ratchet behaviour, not a bug: only `decrypt` discovers `theirPublicKey` and completes the
    /// matching sending chain via the standard DH-ratchet step.
    public static func initReceiver(sharedRootKey: Data, ownKeyPair: KeyPair) -> RatchetState {
        RatchetState(
            dhsPrivateKey: ownKeyPair.privateKey, dhsPublicKey: ownKeyPair.publicKey,
            dhr: nil, rootKey: sharedRootKey, sendChainKey: nil, recvChainKey: nil
        )
    }

    public static func encrypt(state: RatchetState, plaintext: Data, associatedData: Data) throws -> EncryptedMessage {
        guard let chainKey = state.sendChainKey else {
            throw RatchetError("no sending chain yet — must receive at least one message first")
        }
        let (nextChainKey, messageKey) = try kdfChainKey(chainKey: chainKey)
        state.sendChainKey = nextChainKey
        let header = RatchetHeader(dhPub: state.dhsPublicKey, previousChainLength: state.previousChainLength, messageNumber: state.sendCount)
        state.sendCount += 1
        let ciphertext = try aeadSeal(messageKey: messageKey, plaintext: plaintext, aad: associatedData + (try header.encode()))
        return EncryptedMessage(header: header, ciphertext: ciphertext)
    }

    public static func decrypt(state: RatchetState, header: RatchetHeader, ciphertext: Data, associatedData: Data) throws -> Data {
        let fullAd = associatedData + (try header.encode())

        if let messageKey = trySkippedKey(state: state, header: header) {
            return try aeadOpen(messageKey: messageKey, ciphertext: ciphertext, aad: fullAd)
        }

        if state.dhr == nil || header.dhPub != state.dhr {
            try skipMessageKeys(state: state, until: header.previousChainLength)
            try dhRatchetStep(state: state, theirNewPublicKey: header.dhPub)
        }

        try skipMessageKeys(state: state, until: header.messageNumber)
        guard let chainKey = state.recvChainKey else {
            throw RatchetError("receiving chain not established")
        }
        let (nextChainKey, messageKey) = try kdfChainKey(chainKey: chainKey)
        state.recvChainKey = nextChainKey
        state.recvCount += 1
        return try aeadOpen(messageKey: messageKey, ciphertext: ciphertext, aad: fullAd)
    }

    // MARK: - DH ratchet

    private static func dhRatchetStep(state: RatchetState, theirNewPublicKey: Data) throws {
        state.previousChainLength = state.sendCount
        state.sendCount = 0
        state.recvCount = 0
        state.dhr = theirNewPublicKey

        let (rk1, recvChain) = try kdfRootKey(rootKey: state.rootKey, dhOutput: try dh(privateKey: state.dhsPrivateKey, publicKey: theirNewPublicKey))
        state.rootKey = rk1
        state.recvChainKey = recvChain

        let newKeyPair = generateKeyPair()
        state.dhsPrivateKey = newKeyPair.privateKey
        state.dhsPublicKey = newKeyPair.publicKey

        let (rk2, sendChain) = try kdfRootKey(rootKey: state.rootKey, dhOutput: try dh(privateKey: state.dhsPrivateKey, publicKey: theirNewPublicKey))
        state.rootKey = rk2
        state.sendChainKey = sendChain
    }

    // MARK: - Skipped message keys (out-of-order delivery)

    private static func skipMessageKeys(state: RatchetState, until: Int32) throws {
        guard let chainKey = state.recvChainKey else { return } // nothing received on this chain yet
        if until - state.recvCount > maxSkippedKeys {
            throw RatchetError("too many skipped messages (\(until - state.recvCount))")
        }
        var ck = chainKey
        while state.recvCount < until {
            let (nextChainKey, messageKey) = try kdfChainKey(chainKey: ck)
            ck = nextChainKey
            state.skippedKeys[skippedKeyId(dhPub: state.dhr!, n: state.recvCount)] = messageKey
            if state.skippedKeys.count > maxSkippedKeys {
                throw RatchetError("skipped-key cache overflow")
            }
            state.recvCount += 1
        }
        state.recvChainKey = ck
    }

    private static func trySkippedKey(state: RatchetState, header: RatchetHeader) -> Data? {
        let id = skippedKeyId(dhPub: header.dhPub, n: header.messageNumber)
        return state.skippedKeys.removeValue(forKey: id)
    }

    private static func skippedKeyId(dhPub: Data, n: Int32) -> String {
        dhPub.base64EncodedString() + ":" + String(n)
    }

    // MARK: - Primitives

    private static func dh(privateKey: Data, publicKey: Data) throws -> Data {
        let priv = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: privateKey)
        let pub = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: publicKey)
        let secret = try priv.sharedSecretFromKeyAgreement(with: pub)
        return secret.withUnsafeBytes { Data($0) }
    }

    /// KDF_RK: derives a fresh root key + chain key from the running root key and a new DH output.
    private static func kdfRootKey(rootKey: Data, dhOutput: Data) throws -> (rootKey: Data, chainKey: Data) {
        let out = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: dhOutput),
            salt: rootKey,
            info: rootKdfInfo,
            outputByteCount: 64
        ).withUnsafeBytes { Data($0) }
        return (out.subdata(in: 0..<32), out.subdata(in: 32..<64))
    }

    /// KDF_CK: advances a symmetric chain by one step, yielding (nextChainKey, messageKey).
    private static func kdfChainKey(chainKey: Data) throws -> (nextChainKey: Data, messageKey: Data) {
        let nextChainKey = hmac(key: chainKey, data: Data([0x02]))
        let messageKey = hmac(key: chainKey, data: Data([0x01]))
        return (nextChainKey, messageKey)
    }

    private static func hmac(key: Data, data: Data) -> Data {
        let mac = HMAC<SHA256>.authenticationCode(for: data, using: SymmetricKey(data: key))
        return Data(mac)
    }

    // MARK: - AEAD (XChaCha20-Poly1305, via CryptoSwift — see file doc comment)

    private static let xchachaNonceSize = 24
    private static let xchachaTagSize = 16

    private static func aeadSeal(messageKey: Data, plaintext: Data, aad: Data) throws -> Data {
        var nonceBytes = [UInt8](repeating: 0, count: xchachaNonceSize)
        let status = SecRandomCopyBytes(kSecRandomDefault, xchachaNonceSize, &nonceBytes)
        guard status == errSecSuccess else { throw RatchetError("failed to generate nonce") }

        let result = try AEADXChaCha20Poly1305.encrypt(
            [UInt8](plaintext), key: [UInt8](messageKey), iv: nonceBytes, authenticationHeader: [UInt8](aad)
        )
        // Matches Tink's XChaCha20Poly1305 output framing on the Android side: nonce || ciphertext || tag.
        var out = Data(nonceBytes)
        out.append(contentsOf: result.cipherText)
        out.append(contentsOf: result.authenticationTag)
        return out
    }

    private static func aeadOpen(messageKey: Data, ciphertext: Data, aad: Data) throws -> Data {
        guard ciphertext.count >= xchachaNonceSize + xchachaTagSize else {
            throw RatchetError("ciphertext too short")
        }
        let bytes = [UInt8](ciphertext)
        let nonce = Array(bytes[0..<xchachaNonceSize])
        let tagStart = bytes.count - xchachaTagSize
        let inner = Array(bytes[xchachaNonceSize..<tagStart])
        let tag = Array(bytes[tagStart...])

        let result = try AEADXChaCha20Poly1305.decrypt(
            inner, key: [UInt8](messageKey), iv: nonce, authenticationHeader: [UInt8](aad), authenticationTag: tag
        )
        guard result.success else { throw RatchetError("AEAD authentication failed") }
        return Data(result.plainText)
    }
}
